lib/signalfx/signalflow/computation.rb

```diff
@@ -0,0 +1,74 @@
+# Represents a SignalFlow computation.  A computation can have a channel
+# associated with it, but not necessarily (it could have been detached while
+# the computation is still running).  New instances should only be created by
+# the client.
+class Computation
+  attr_accessor :handle
+  attr_accessor :channel
+
+  def initialize(handle, attach_func, stop_func)
+    @handle = handle
+    @channel = nil
+    @attach_func = attach_func
+    @stop_func = stop_func
+  end
+
+  def channel=(channel)
+    @channel = channel
+
+    # This is done in a thread because of the circular nature of channels and
+    # computations in order to avoid trying to grab the same lock recursively
+    # in the channel.  The channel is set in response to a message on the
+    # channel and we can't iterate messages while processing a message.
+    Thread.new do
```

Comment:
I don't like all this with the thread and all the locking you have to do in `Channel`. Given my above comment, since a `Computation` object is always attached it will "see" all messages from the beginning of the stream by design. Which means you don't need to have this message replay logic.

Maybe part of the complexity you've brought upon yourself is that you support multiple "blocks" receiving messages from a given channel, but really only the computation should be getting messages from the channel. Then, at the `Computation` object layer, we can do some smarter things if we want to.

Reply:
I will simplify the channel logic and make everything funnel through the computation.